### ISSUE TITLE
span class 'sr-only' should be ignored

### ISSUE THREAD

**Issue by Commodore64user** [koreader/koreader#14124]
### Device

Other (please specify in additional information)

### KOReader version

2025.04

### Issue description

Not really a bug, but screen reader information should not be rendered by us.

```html
<header>
</header><p class="h1" aria-hidden="true">11</p>
<h1 epub:type="title" class="h2a"><a href="part0002.xhtml#rch11"><span class="sr-only">Chapter 11, </span><em>Late-September Shootings</em></a></h1>
```
here is a screenshot of koreader and calibre rendering the same text. As far as I can see, we are the only ones doing so.

<p align="center">
<img width="1036" height="293" alt="Image" src="https://github.com/user-attachments/assets/09ed0724-1c24-476b-a984-a1b2f0993392" />
</p> 

### Steps to reproduce

2. Click on '...'
3. Scroll down to '...'

### Log

```shell

```

**Comment by Frenzie** [comment]
I'm not convinced by that as written in the title. Possibly as a style tweak.

calibre will simply be applying something like `.sr-only {display: none}` from the book itself. The question is what the precise CSS involved is.

Maybe you could share it with [ScrambleEbook](https://www.mobileread.com/forums/showthread.php?t=267998)?

**Comment by poire-z** [comment]
Yes, it's not for us to decide that `.sr-only` means "screen reader only" and should be `display:none`.
There must be something in the publisher stylesheets that says that. It may be under some conditions like `@media` that we don't catch/support. So look for that.

**Comment by Commodore64user** [comment]
I don't think is a matter of "for us do decide", those aria-labels and sr-only seem to be used consistently for accessibility.

this is what's under the css style for it

```css
.sr-only 
{
clip:rect(1px, 1px, 1px, 1px);
position:absolute !important;
white-space:nowrap;
height:1px;
width:1px;
overflow:visible;
}
```

[scrambled.epub.txt](https://github.com/user-attachments/files/21558026/scrambled.epub.txt)

**Comment by poire-z** [comment]
They may, but we don't support anything "accessibility" explicitely.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-hidden doesn't say we should not show it, just that:
> Adding aria-hidden="true" to an element removes that element and all of its children from the accessibility tree. This can improve the experience for assistive technology users by hiding:
    Purely decorative content, such as icons or images [...]

So, we'd rather not remove decorative content - if it's there, it's to be shown on screen :)

As for the `.sr-only` class, again, it's just a class without any semantic associated. We shouldn't do anything because the class name sounds like "screen reader only". We render HTML according to the associated CSS, period.

> this is what's under the css style for it
> 
> .sr-only 
> {
> clip:rect(1px, 1px, 1px, 1px);
> position:absolute !important;
> white-space:nowrap;
> height:1px;
> width:1px;
> overflow:visible;
> }

Well, lot's of CSS tricks - that we don't support - to have the content not hidden (no `display:none`, no `visibility:hidden`, possibly so it's not considered hidden by screen readers), but not visible/rendered (the `clip`, but why `overflow:visible` ?).

I think it's really a job for a user style tweak, if you have lots of books using this.
I believe I've never met a book needing this.
If it happens to be really common and consistent, it could become a shipped style tweaks, possibly named "Remove classic accessibility elements". You could put it as default, but at least, you could disable it to reveal what may have been erroneously hidden. But you'll have to tune it for quite some time to be sure it does not too much harm.

**Comment by Frenzie** [comment]
> I don't think is a matter of "for us do decide", those aria-labels and sr-only seem to be used consistently for accessibility.

`.sr-only` is an accessibility pattern that occurs, but that it's not for us to decide is exactly what I'm saying. Incidentally ARIA is different; that's a defined standard, but not of any particular relevance to us unless someone put in a humongous amount of effort for screenreader support. (Which wouldn't be a bad thing I'm sure but it seems a bit unlikely to happen. ;-) I suppose you could write a "show ARIA labels" style tweak or something.

Note that Bootstrap 5 no longer has `.sr-only` like Bootstrap 3, but instead has switched to `.visually-hidden` and `.visually-hidden-focusable`. Its CSS is [similar](https://github.com/twbs/bootstrap/blob/b850fcbe49f940008b942b381935f1d2b4c84216/scss/mixins/_visually-hidden.scss#L9-L16) to that above.

```css
  width: 1px !important;
  height: 1px !important;
  padding: 0 !important;
  margin: -1px !important; // Fix for https://github.com/twbs/bootstrap/issues/25686
  overflow: hidden !important;
  clip: rect(0, 0, 0, 0) !important;
  white-space: nowrap !important;
  border: 0 !important;
```

**Comment by Frenzie** [comment]
PS MuPDF and Okular also don't fully understand that CSS.

<img width="562" height="854" alt="Image" src="https://github.com/user-attachments/assets/9c58fd5b-68fe-47e0-b28f-256d24a39f47" />

<img width="3200" height="1800" alt="Image" src="https://github.com/user-attachments/assets/d20bc8af-db53-4058-a5fb-1c76909daeb8" />

**Comment by Commodore64user** [comment]
We have no support for that kind of accessibility [feature], so why show it? Even on my kindle keyboard, the "chapter x" is not displayed, only when using text to speech, do you hear the "chapter x". Shouldn't we be doing the same?

**Comment by poire-z** [comment]
> We have no support for that kind of accessibility, so why show it? 

About the aria:hiden=true, because nothing (not even the specs) says we should not show it.
If it's about the CSS, because we don't support the properties used (may be Kindle does). If we were to not show it, it would be by supporting these CSS properties in a generic way, not by forcing (via our user agent stylesheet, of programatically) `display:none`. But you are free to do it via a user style tweak (or a shipped style tweaks after you have battle tested it).
And these properties `width/height` on inline elements, `clip` and `overflow`, are hard to implement, and not really useful and hardly seen in books (they are mostly for fancy web site layouts).

**Comment by Frenzie** [comment]
Perhaps clip(0,0,0,0) could be a special case, more or less a synonym of display none? Probably too hacky, but then again so is the CSS.

**Comment by Commodore64user** [comment]
This one is (1,1,1,1) though ;)

I think @poire-z should consider it

**Comment by samimaattaCelia** [comment]
Hello! This is off-topic, but I couldn't find this information elsewhere and here accessibility is somewhat discussed.

> -- Incidentally ARIA is different; that's a defined standard, but not of any particular relevance to us unless someone put in a humongous amount of effort for screenreader support. (Which wouldn't be a bad thing I'm sure but it seems a bit unlikely to happen. ;-) --

Does this quote from @Frenzie mean that KOReader has no support for assistive technologies or even screen readers?

**On topic**

I also didn't quite understand if you were considering using `display: none;` in place of `.sr-only` CSS class.  You should know that `display: none;` also hides information from screen readers. It can't be used interchangeably with the CSS class. Source: [WebAIM: CSS in Action - Invisible Content Just for Screen Reader Users](https://webaim.org/techniques/css/invisiblecontent/)

**On MathML support**

I got interested in KOReader, because it supports visually rendering MathML quite well in mobile (better than I have encountered elsewhere), but does it relay MathML to screen readers?

**Comment by Frenzie** [comment]
Yes, it's an app for ereaders (=low-power screen-only devices) first. The rendering engine is also custom and much lighter for that reason, not an off-the-shelf browser engine. If you're looking for software with screenreader support I'm sure there must be a reasonable selection desktop-first or Android-first apps, most likely based on a regular browser engine.

> rendering MathML quite well in mobile (better than I have encountered elsewhere)

I find that somewhat surprising. I wouldn't really expect anything based on a browser engine to do particularly better or worse. Very nicely done then @poire-z :-)
